src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,397 @@
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Text.Operations;
+using Microsoft.VisualStudio.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.CompleteStatement
+{
+    /// <summary>
+    /// When user types <c>;</c> in a statement, closing delimiters and semicolon are added and caret is placed after the semicolon
+    /// </summary>
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [Name(nameof(CompleteStatementCommandHandler))]
+    [Order(After = PredefinedCommandHandlerNames.Completion)]
+    internal sealed class CompleteStatementCommandHandler : IChainedCommandHandler<TypeCharCommandArgs>
+    {
+        private readonly ITextUndoHistoryRegistry _undoHistoryRegistry;
+        private readonly IEditorOperationsFactoryService _editorOperationsFactoryService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CompleteStatementCommandHandler(ITextUndoHistoryRegistry undoHistoryRegistry, IEditorOperationsFactoryService editorOperationsFactoryService)
+        {
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        public string DisplayName => CSharpEditorResources.Complete_statement_on_semicolon;
+
+        public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            BeforeExecuteCommand(args, executionContext);
+            nextCommandHandler();
+        }
+
+        private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';')
+            {
+                return;
+            }
+
+            var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caret.HasValue)
+            {
+                return;
+            }
+
+            var isCaretAtEndOfLine = ((SnapshotPoint)caret).Position == ((SnapshotPoint)caret).GetContainingLine().End;
+            var document = caret.Value.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+      
+            {
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Value.Position;
+
+            var token = root.FindToken(caretPosition);
+
+            var currentNode = token.Parent;
+
+            // If cursor is right before an opening delimiter, start with node outside of delimiters. 
+            // This covers cases like `obj.ToString$()`, where `token` references `(` but the caret isn't actually 
+            // inside the argument list.
```

Comment:
i don't really understand the comment.  we allow the user to type `;` before a `(arglist)`?  Is that an important case to care about?

Reply:
@chborl The comment here may benefit from the following addition:

> If cursor is right before an opening delimiter, start with node outside of delimiters**, since analysis starting with a node containing delimiters assumes the caret is placed inside those delimiters**.

Reply:
I'm confused by that though.  This code isnt' about us bailing out, this is about us handling the case where we're right before the open paren.  If they "intended to precede a previously-existing statement that started with a parenthesized expression" then why are absorbing that subsequent statement into the analysis?

basically, i don't get why we have this very special logic, but we don't, for example, handle ';' before many other types of constructs.

Reply:
➡️ @CyrusNajmabadi This code is handling the case where `FindToken` returns a `(` token, without distinguishing between the case where the caret is inside or outside the parentheses. The analysis assumes that if we are starting at a parenthesized expression, we are doing so because we are *inside* the parentheses. If `FindToken` gives us a `(` but we are not inside the parentheses, we force the analysis to start outside the parenthesized expression because the caret is outside the parentheses. The comment has already been updated as I recommended earlier. I'm going to mark this as resolved given the current comments and presence of tests covering this case (the latter to highlight importance of the handling in a case where a user mistakenly removes the special handling).